integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.logp(Level.FINE, this.getClass().getName(), "start",
+                        "Created new Association ({0}) for connection ({1}) in state ACTIVE", new Object[] {a, c});
+        }
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "start", a);
+        }
+        return a;
+    }
+
+    @Override // XAResource
+    public void end(Xid xid, int flags) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "end", new Object[] {xid, flagsToString(flags)});
+        }
+        requireNonNullXid(xid);
+
+        BiFunction<Xid, Association, Association> remappingFunction;
+        switch (flags) {
+        case TMFAIL:
+        case TMSUCCESS:
+            remappingFunction = LocalXAResource::activeToIdle;
+            break;
+        case TMSUSPEND:
+            remappingFunction = LocalXAResource::suspend;
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("xid: " + xid + "; flags: " + flagsToString(flags)));
+        }
+
+        // Any XAException thrown can have any error code. The transaction will be marked as rollback only. See
+        // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L978-L992.
+        this.computeAssociation(XARoutine.END, xid, remappingFunction);
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "end");
+        }
+    }
+
+    @Override // XAResource
+    public int prepare(Xid xid) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "prepare", xid);
+        }
+        requireNonNullXid(xid);
+
+        // Any XAException thrown can have basically any error code. See
+        // https://github.com/jbosstm/narayana/blob/c5f02d07edb34964b64341974ab689ea44536603/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L227-L261.
+        Object association =
+            this.computeAssociation(XARoutine.PREPARE,
+                                    xid,
+                                    EnumSet.of(Association.BranchState.IDLE),
+                                    LocalXAResource::prepare,
+                                    false); // don't remove association on error
+
+        int returnValue = association == null ? XA_RDONLY : XA_OK;
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "prepare", returnValue);
+        }
+        return returnValue;
+    }
+
+    @Override // XAResource
+    public void commit(Xid xid, boolean onePhase) throws XAException {
```

Comment:
This method is called by the transaction manager during the two-phase completion stage, after the `start`/`end` sequence has occurred, and possibly after a `prepare` invocation (if `onePhase` is false).  At this point the resource manager (represented by the `Connection` in our case) must commit its work, or the XA protocol will be deemed to have been violated.